File: src/answers.c
#include <gmodule.h>
#include <stdlib.h>

#include "answers.h"


/*
Estrutura dos posts respostas.
*/
typedef struct answers {
  long user_id;
  long answer_id;
  long parent_id; // a que pergunta pertence
  int score;
  int comment_count; //número de comentários recebidos pela resposta (query 10)
} answers;

/*
Função que devolve o tamanho da estrutura answers.
@returns int - Tamanho da estrutura answers.
*/
int sizeAnswers() {
    return sizeof(struct answers);
}

/*
Função que devolve o id do user que deu a resposta
@param a Apontador para answers.
@returns long - Título e utilizador.
*/
long getAUserId(Answers a) {
    return a->user_id;
}

/*
Função que estabelece o id do user que deu determinada resposta
@param a Apontador para answers.
@param id Identificador do user.
*/
void setAUserId(Answers a, long id) {
    a->user_id = id;
}

/*
Função que devolve o id da resposta
@param a Apontador para answers.
@returns long - Identificador da resposta.
*/
long getAnswerId(Answers a) {
    return a->answer_id;
}

/*
Função que estabelece o id da resposta
@param a Apontador para answers.
@param id Identificador da resposta.
*/
void setAnswerId(Answers a, long id) {
    a->answer_id = id;
}

/*
Função que devolve o id da pergunta a que a resposta se refere
@param a Apontador para answers.
@returns long - Identificador da pergunta a que a resposta está associada.
*/
long getParentId(Answers a) {
    return a->parent_id;
}

/*
Função que estabelece o id da pergunta a que a resposta se refere
@param a Apontador para answers.
@param id Identificador da pergunta a que a resposta está associada.
*/
void setParentId(Answers a, long id) {
    a->parent_id = id;
}

/*
Função que devolve o score de uma resposta
@param a Apontador para answers.
@returns int - Score.
*/
int getScore(Answers a) {
    return a->score;
}

/*
Função que estabelece o score de determinada resposta
@param a Apontador para answers.
@param s score.
*/
void setScore(Answers a, int s) {
    a->score = s;
}

/*
Função que devolve o número de comentários de uma resposta
@param a Apontador para answers.
@returns int - Número de Comentários.
*/
int getCommentCount(Answers a) {
    return a->comment_count;
}

/*
Função que estabelece o número de comentários de uma resposta
@param a Apontador para answers.
@param c Número de comentários.
*/
void setCommentCount(Answers a, int c) {
    a->comment_count = c;
}

/*
Função que compara os votos de duas respostas
(query 6)
@param a1 Apontador para apontador para answers.
@param a2 Apontador para apontador para answers.
@returns gint - positivo se a1 < a2 e negativo caso contrário.
*/
static gint sortVotes(Answers * a1, Answers * a2) {
  int votes_a1 = 0, votes_a2 = 0;
  votes_a1 = getScore(*a1);
  votes_a2 = getScore(*a2);

    if(votes_a1 < votes_a2)
      return 1;
    else
      return -1;
}


/*
Função que ordena os votos das respostas por ordem decrescente do número de votos
(query 6)
@param total_answers Apontador para um GPtrArray.
*/
void sortAnswersVotes(GPtrArray * total_answers) {
  g_ptr_array_sort(total_answers, (GCompareFunc)sortVotes);
}

/*
Função que liberta o apontador para a estrutura answers.
@param a Apontador para answers.
*/
void cleanAnswer(Answers a) {
    free(a);
}

File: proj-java/src/main/java/common/Answers.java
package common;


/**
 * Estrutura de dados de Posts tipo 2
 *
 * @author A81047
 * @author A34900
 * @author A82339
 * @version 20180519
 */

public class Answers extends Posts {
    private long parent_id; // a que pergunta pertence
    private int score;
    private int comment_count; //número de comentários recebidos pela resposta (query 10)

    /**
     * Construtor por omissão de Answers.
     */
    public Answers() {
        super();
        parent_id = 0;
        score = 0;
        comment_count = 0;
    }

    /**
     * Construtor parametrizado de Users.
     *
     * @param user_id Identificador do user
     * @param post_id Identificador do post
     * @param parent_id Identificador da pergunta a que se refere
     * @param score Score da resposta
     * @param comment_count Número de comentários
     */
    public Answers(long user_id, long post_id, long parent_id, int score, int comment_count) {
        super(user_id, post_id);
        this.parent_id = parent_id;
        this.score = score;
        this.comment_count = comment_count;
    }

    /**
     * Construtor de Cópia.
     *
     * @param umCI Contribuinte Individual a ser replicado
     */
    public Answers(Answers a) {
        super(a);
        this.parent_id = a.getParentId();
        this.score = a.getScore();
        this.comment_count = a.getCommentCount();
    }

    /**
     * Função que devolve o id da pergunta a que a resposta pertence.
     *
     * @returns long - Identificador da pergunta a que a resposta pertence.
     */
    public long getParentId() {
        return parent_id;
    }

    /**
     * Função que estabelece o id da pergunta a que a resposta pertence.
     *
     * @param id Identificador da pergunta a que a resposta pertence.
     */
    public void setParentId(long id) {
        parent_id = id;
    }

    /**
     * Função que devolve o score da resposta
     *
     * @returns int - score
     */
    public int getScore() {
        return score;
    }

    /**
     * Função que estabelece o score da resposta
     *
     * @param score Score da resposta
     */
    public void setScore(int score) {
        this.score = score;
    }

    /**
     * Função que devolve o número de comentários da resposta.
     *
     * @returns int - número de comentários
     */
    public int getCommentCount() {
        return comment_count;
    }

    /**
     * Função que estabelece o número de comentários da resposta.
     *
     * @param comment_count o número de comentários da resposta.
     */
    public void setCommentCount(int comment_count) {
        this.comment_count = comment_count;
    }

    public Answers clone() {
         return new Answers(this);
    }
    
    /**
     * Método que determina se duas respostas são iguais.
     *
     * @param object Objecto a ser usado como termo de comparação.
     *
     * @return Boolean indicando se os dois objetos são iguais
     */
    public boolean equals(Object object) {
        if(this == object)
            return true;

        if(object == null || (this.getClass() != object.getClass()))
            return false;

        Answers resposta = (Answers) object;
        return  (super.equals(resposta) && parent_id == resposta.getParentId() 
                 && score == resposta.getScore() && comment_count == resposta.getCommentCount());

    }
    
     /**
     * Método que devolve a representação em String de Answers.
     *
     * @return String que representa uma resposta.
     */
    public String toString() {

        return "Answers{" +
                super.toString() +
                ", ParentId = " + parent_id +
                ", Score = " + score +
                ", CommentCount = " + comment_count +
                '}';
    }
}
